### PR TITLE
give more time to workers to finish the process

### PR DESCRIPTION
**Pull request by byara**
This PR gives more time to the workers to finish before ending them.

### CODE REVIEW

**Comment by liangchunn** [comment]
I think... we need a better way 😛

File: index.js [hunk 1]
```diff
@@ -114,6 +114,15 @@ function run(configPath, options, callback) {
         workerFarm.end(workers);
     };
 
+    function injectTimeout(cb, time){
+        setTimeout(cb, time);
+    }
+
+    function finallCallback(){
```
Comment (frvge) [inline index.js]: final

File: index.js [hunk 2]
```diff
@@ -149,6 +157,8 @@ function run(configPath, options, callback) {
     return farmPromise;
 }
 
+
```
Comment (frvge) [inline index.js]: whitespace

**Comment by frvge** [comment]
@efegurkan , what do you think?

**Comment by efegurkan** [comment]
IMO, this is a hack that should be solved better. You should find a way to add this timeout as a flag, config etc. otherwise it will be a breaking change.

File: index.js [hunk 3]
```diff
@@ -138,9 +147,8 @@ function run(configPath, options, callback) {
         if(results.length) {
             return results;
         }
-    }).finally(function () {
-        workerFarm.end(workers);
-        process.removeListener('SIGINT', shutdownCallback);
+    }).finally(function() {
+        injectTimeout(finallCallback, 10000);
```
Comment (pago) [inline index.js]: Does a smaller number work as well? Is there any real signal that we can hook into to find out the right timeout?

**Comment by liangchunn** [comment]
@pago 
To be honest, I _think_ this problem is related to how `webpack` and `webpack-bundle-analyzer` interacts, and this PR will just 'work around' the issues that @byara and some other people have encountered.

From what I've seen in the source of `webpack-bundle-analyzer`, it defers the job of actually analyzing and emitting the stats by utilizing `setImmediate(() => {})` [here](https://github.com/webpack-contrib/webpack-bundle-analyzer/blob/master/src/BundleAnalyzerPlugin.js#L58-L60), and then implicitly returns.

That said, webpack actually doesn't know that there's enqueued operations, since `webpack-bundle-analyzer` only does the transformations _after_ I/O operations are done, which at the same time, `webpack` might have already declared it's job done, and calling it's `done` callback, and on the `parallel-webpack` side terminating the worker farm.

This 'hack' works because the farm termination is effectively delayed, and the underlying processes are essentially allowed to do whatever they want (and yes, probably executing the enqueued operations that `webpack-bundle-analyzer` has set). 

I might do a little bit of research/test into this again since it's pretty interesting to me why `webpack-bundle-analyzer` uses `setImmediate` rather than a synchronous call.

File: index.js (hunk 3 above)
Comment (byara) [reply index.js]: I believe getting a result with any number is not guaranteed. we can try smaller numbers though.
For a signal to hook into, I think we should use the new plugin API. The `done` hook would be the best choice, I presume.

Comment (liangchunn) [reply index.js]: This plugin is already listening to the done hook. The issue is that webpack-bundle-analyzer is set up wrongly internally, and only executing after the done hook is called.

Comment (byara) [reply index.js]: @pago I checked different values for the timeout and it seems it depends on the size of the stats file. I'll try to add the timeout as a flag or config.

File: bin/run.js [hunk 4]
```diff
@@ -47,6 +47,7 @@ if(argv.version) {
             exclude: argv['display-exclude'],
             colors: argv['colors'],
             stats: argv['stats'],
+            timeout: argv['timeout'],
```
Comment (pago) [inline bin/run.js]: Could you add it to the schema as well?
https://github.com/trivago/parallel-webpack/blob/master/schema.json

**Comment by pago** [comment]
@byara @liangchunn Last thing I'm still wondering about is the name of the flag. `timeout` indicates to me that the build will stop after that time but here its used to describe a mandatory delay.
How about `keep-alive-after-finish` or something similar?

**Comment by byara** [comment]
> @byara @liangchunn Last thing I'm still wondering about is the name of the flag. `timeout` indicates to me that the build will stop after that time but here its used to describe a mandatory delay.
> How about `keep-alive-after-finish` or something similar?

absolutely right. I have no problem with that. If we cannot think of something shorter, I will make the changes according to `keep-alive-after-finish`.

**Pull request merged**